### DISCUSSION TITLE
`readFragment` type definition: handling array with null/undefined

### DISCUSSION THREAD

**Discussion by 3ru** [0no-co/gql.tada#337] [Q&A]
I have a question regarding the type definition of the `readFragment` function. I noticed that there are no specific overloads for handling arrays that might be null or undefined. 

Currently, the type definition includes:

https://github.com/0no-co/gql.tada/blob/010a4f32e6564430e68daa85b29e87cfdacdea8a/src/api.ts#L554-L557

However, it doesn't include overloads for the following cases:
```typescript
declare function readFragment<const Document extends FragmentShape>(
  _document: Document,
  fragment: readonly resultOrFragmentOf<Document>[] | null
): readonly ResultOf<Document>[] | null;

declare function readFragment<const Document extends FragmentShape>(
  _document: Document,
  fragment: readonly resultOrFragmentOf<Document>[] | undefined
): readonly ResultOf<Document>[] | undefined;

declare function readFragment<const Document extends FragmentShape>(
  _document: Document,
  fragment: readonly resultOrFragmentOf<Document>[] | null | undefined
): readonly ResultOf<Document>[] | null | undefined;
```

I'm wondering if this is the expected behavior or if these cases were intentionally omitted.

For context, I'm in the process of migrating from `@graphql-codegen` to `gql.tada`, and I encountered type errors due to this [difference ](https://github.com/dotansimha/graphql-code-generator/blob/16917691adc00261a3eccd1b3aa0f7d01652ef5e/packages/presets/client/src/fragment-masking-plugin.ts#L71)in type definitions.

**Comment by kitten** [comment]
This is really a maintenance problem.

We have some more use-cases pending for `readFragment`, but currently, every added case exponentially increase the amount of overloads, of course. So, for a new case we're considering, we'd have to double all overloads. If we add some more then this quickly gets out of hand.

The underlying problem is that while we've tried non-overload patterns, they're simply not as performant for TypeScript to evaluate for this specific case, or don't represent the API we want. So, we'll need some time to find some alternatives and/or also check the implications of adding a huge amount of overloads (in terms of inference performance as well) before adding more cases.

This said, just to point this out. These are all helpers. There's nothing stopping you from calling `readFragment` on individual items.
